server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/Worker.scala

```diff
@@ -836,16 +889,91 @@ private[deploy] class Worker(
     }
   }
 
-  override def handleOpenStream(
-      shuffleKey: String, fileName: String): FileInfo = {
-    logDebug(s"Open file $fileName for $shuffleKey")
+  val newOpenStreamMapFunc =
+    new java.util.function.BiFunction[String, ConcurrentSet[String], ConcurrentSet[String]]() {
+      override def apply(key: String, u: ConcurrentSet[String]): ConcurrentSet[String] = {
+        if (u == null) {
+          new ConcurrentSet[String]
+        } else {
+          u
+        }
+      }
+    }
+
+  override def handleOpenStream(shuffleKey: String, fileName: String, startMapIndex: Int,
+    endMapIndex: Int): FileInfo = {
+    logInfo(s"Open file $fileName for $shuffleKey with" +
+      s" startMapIndex ${startMapIndex} endMapIndex ${endMapIndex}")
     // find FileWriter responsible for the data
     val fileWriter = localStorageManager.getWriter(shuffleKey, fileName)
     if (fileWriter eq null) {
       logWarning(s"File $fileName for $shuffleKey was not found!")
       return null
     }
-    new FileInfo(fileWriter.getFile, fileWriter.getChunkOffsets, fileWriter.getFileLength)
+
+    if (endMapIndex != Int.MaxValue) {
+      val shuffleFileId = shuffleKey + fileName;
+      val shuffleBoundSortedShuffleFiles = sortedShuffleFiles.compute(shuffleKey,
+        newOpenStreamMapFunc)
+      val shuffleBoundSortingShuffleFiles = sortingShuffleFiles.compute(shuffleKey,
+        newOpenStreamMapFunc)
+
+      if (!shuffleBoundSortedShuffleFiles.contains(shuffleFileId)) {
+        if (!shuffleBoundSortingShuffleFiles.contains(shuffleFileId)) {
+          logInfo(s"[Worker] sorting file ${shuffleFileId}")
+          shuffleBoundSortingShuffleFiles.add(shuffleFileId)
+          val sortSuccess = new AtomicBoolean(false)
+
+          val sortTask = new PartitionFileSorter(fileWriter.getFile, readChunkSize,
```

Comment:
There should be one PartitionFilesSorter instance in Worker, and refine apis of it